index.js

```diff
@@ -1,189 +1,317 @@
+ * paths. Use in conjunction with `internalGetPathValue`.
  *
- *  var parsed = parsePath('myobject.property.subprop');
+ *      var parsed = parsePath('myobject.property.subprop');
  *
  * ### Paths:
  *
- * * Can be as near infinitely deep and nested
+ * * Can be infinitely deep and nested.
  * * Arrays are also valid using the formal `myobject.document[3].property`.
+ * * Literal dots and brackets (not delimiter) must be backslash-escaped.
  *
  * @param {String} path
- * @returns {Array} parsed
+ * @returns {Object} parsed
+ * @api private
  */
 
-exports.parse = function(path) {
-  var str = (path || '').replace(/\[/g, '.[');
+function parsePath(path) {
+  var str = path.replace(/([^\\])\[/g, '$1.[');
   var parts = str.match(/(\\\.|[^.]+?)+/g);
-  var re = /\[(\d+)\]$/;
-  var ret = [];
-  var mArr = null;
-
-  for (var i = 0, len = parts.length; i < len; i++) {
-    mArr = re.exec(parts[i]);
-    ret.push(mArr ? { i: parseFloat(mArr[1]) } : { p: parts[i] });
-  }
+  return parts.map(function mapMatches(value) {
+    var regexp = /^\[(\d+)\]$/;
+    var mArr = regexp.exec(value);
+    var parsed = null;
+    if (mArr) {
+      parsed = { i: parseFloat(mArr[1]) };
+    } else {
+      parsed = { p: value.replace(/\\([.\[\]])/g, '$1') };
+    }
 
-  return ret;
-};
+    return parsed;
+  });
+}
 
-/*!
- * Companion function for `parsePath` that returns
+/* !
+ * ## internalGetPathValue(parsed, obj)
+ *
+ * Helper companion function for `.parsePath` that returns
  * the value located at the parsed address.
  *
- *  var value = getPathValue(parsed, obj);
+ *      var value = getPathValue(parsed, obj);
  *
  * @param {Object} parsed definition from `parsePath`.
  * @param {Object} object to search against
+ * @param {Number} object to search against
  * @returns {Object|Undefined} value
+ * @api private
  */
 
-function getPathValue(parsed, obj) {
-  var tmp = obj;
-  var res;
+function internalGetPathValue(parsed, obj, pathDepth) {
+  var temporaryValue = obj;
+  var res = null;
+  pathDepth = (typeof pathDepth === 'undefined' ? parsed.length : pathDepth);
 
-  for (var i = 0, l = parsed.length; i < l; i++) {
+  for (var i = 0, pathPart = pathDepth; i < pathPart; i++) {
     var part = parsed[i];
-    if (tmp) {
-      if (defined(part.p)) tmp = tmp[part.p];
-      else if (defined(part.i)) tmp = tmp[part.i];
-      if (i == (l - 1)) res = tmp;
-    } else {
-      res = undefined;
+    if (temporaryValue) {
+      if (typeof part.p === 'undefined') {
+        temporaryValue = temporaryValue[part.i];
+      } else {
+        temporaryValue = temporaryValue[part.p];
+      }
+
+      if (i === (pathPart - 1)) {
+        res = temporaryValue;
+      }
     }
   }
 
   return res;
-};
+}
+
+/* !
+ * ## defined(val)
+
+ * Check if `val` is defined.
+ *
+ * @param {Mixed} val
+ * @returns {Boolean} `true` if defined
+ * @api private
+ */
+
+function defined(val) {
+  return !(!val && typeof val === 'undefined');
```

Comment:
Better than a comment - if it makes a difference, prove it with a benchmark that we can re-run (like the ones in type-detect). If it doesn't make a (significant) difference, then ditch it.